Implement AUDIO header

TheNailDev opened the pull request:
Format 1.1.0 introduced the mandatory AUDIO header as a replacement for the MP3 header.

USDX now requires songs with VERSION >= 1.1.0 to include the AUDIO header and reads the audio file from this header.
The MP3 header is ignored for songs with VERSION >= 1.1.0.

When saving songs, the AUDIO and MP3 headers are written depending on the VERSION of the song, as shown in the table below:

| VERSION | MP3 | AUDIO |
|--------|--------|--------|
| < 1.1.0 | :heavy_check_mark:   | :x:  |
| >= 1.1.0 / < 2.0.0 | :heavy_check_mark: | :heavy_check_mark: |
| >= 2.0.0[^1] | :x: | :heavy_check_mark: |

[^1]: VERSION >= 2.0.0 is currently not supported by USDX. This is just future proofing

barbeque-squared commented on src/base/USong.pas:
```diff
@@ -1061,20 +1083,35 @@ function TSong.ReadTXTHeader(SongFile: TTextFileStream; ReadCustomTags: Boolean)
+        // If AUDIO is present MP3 should be ignored
+        if TagMap.IndexOf('MP3') > -1 then
+        begin
+          Log.LogInfo('The AUDIO header overwrites the MP3 header in file ' + FullFileName, 'TSong.ReadTXTHeader');
```
Isn't this going to generate a lot of logspam? I'm writing this comment without looking at the spec, I can look at it tomorrow. But basically, if the spec already defines something about which tag is more important, I don't think we need to log about it as well.

bohning replied:
I think it's good to log this. This helps users pull their song collections to newer versions if they want to.

TheNailDev replied:
I thought it to be good to log this as AUDIO needs to be used, even if the file defined in AUDIO is invalid and the file defined in MP3 is valid. Otherwise some users might get confused why the song is not working, even if the MP3 tag points to a valid file.
Additionally it uses the info log level, which should afaik not logged by default, such that it would not actually spam the logs. (Correct me if I'm wrong regarding the default log levels. I have not yet looked too deep into how USDX actually handles logs.)

barbeque-squared replied:
USDX logs everything regardless of log level to Error.log. Which is probably a problem on its own, but given the explanation it's fine to log it.

barbeque-squared commented on src/base/USong.pas:
```diff
@@ -911,6 +911,28 @@ function TSong.ReadTXTHeader(SongFile: TTextFileStream; ReadCustomTags: Boolean)
   end;
 
+  {**
+   * Updates the Audio file of the song (MP3 field) to a file witn given filename.
```
Comment still references MP3 field. Also a typo in `witn` (which might have always been there).

But can't we make an internal function that takes _two_ parameters, namely the filename and which field it needs to set)? Then the CheckAndSet functions with one parameter are essentially oneliners.

TheNailDev replied:
The internal field of the `TSong` class is still called `Mp3`. Naturally we could think about renaming this field in the code too, but I didn't know if there would be any unexpected errors resulting from renaming the field.

I don't know if a general CheckAndSet function makes too much sense, as the checking of the input and setting of the Done flags differs depending on what field to set. (I only created the function so that I don't have to repeat the same code for the AUDIO header and MP3 header.)

bohning replied:
Would be nice to rename the field as well, but can't comment on the implications/side effects.

barbeque-squared commented on src/base/UFiles.pas:
```diff
@@ -149,7 +149,11 @@ function SaveSong(const Song: TSong; const Tracks: array of TLines; const Name:
-      SongFile.WriteLine('#MP3:' + EncodeToken(Song.Mp3.ToUTF8));
+      if Song.FormatVersion.MinVersion(1,1,0) then
+        SongFile.WriteLine('#AUDIO:' + EncodeToken(Song.Mp3.ToUTF8));
+      if Song.FormatVersion.MaxVersion(2,0,0) then
```
See other comment about logspam. If it's actually still valid spec then I have less of an issue with this.

TheNailDev replied:
According to spec both MP3 and AUDIO are required headers. (AUDIO introduced in 1.1.0; MP3 removed in 2.0.0)
Saving both MP3 and AUDIO tags is done as a compatibility measure for software only supporting the MP3 header.

barbeque-squared replied:
Makes sense to write both then, resolving this

barbeque-squared reviewed (commented):
Done a quick initial pass. Is this the only change in 1.1.0 or is there more? The last thing I want to do is half-support something now that people are actually putting effort into standardizing it. Because people _will_ hack around it and/or report bugs because we're only implementing it partially (and then comes the hotfix but it misses some edge case and now we're stuck with n+1 things for the next 20 years)

Not sure if I'll get to really dig into the spec until the weekend, hence this quick review.

EDIT: _personally_ I'm not really against partial support, especially if it's an added tag like this, but if it also for example demands that `#START` and `#END` use sane values, then this is not getting in in its current state. It's a bit of a chicken-egg problem, especially with bigger spec changes it'll be good to split them up across multiple PR's, but ideally it's still all in a single release. Like I said, I need to actually research the spec soon.

TheNailDev commented:
> Is this the only change in 1.1.0 or is there more?

There are more changes for 1.1.0 (see #921):
New headers in 1.1.0 are AUDIO, VOCALS, INSTRUMENTAL, TAGS and PROVIDEDBY.
Additionally 1.1.0 also changes GENRE, CREATOR, EDITION, LANGUAGE and TAGS to multi-valued headers.

IMO support for format 1.1.0 can be classified in distinct sub-features which can be implemented independently:

- Implementation of AUDIO header (this PR)
- Implementation of VOCALS and INSTRUMENTAL (use #850 as base)
- Implementation of the multi valued headers GENRE, CREATOR, EDITION, LANGUAGE and (new header) TAGS
- I don't see a reason to specifically implement PROVIDEDBY, as it is mostly just an information, from which song hosting service (USDB, spanish site, etc.) the song was downloaded. I don't think this information is relevant in USDX. (Naturally we could implement filtering by the song providers, but is that really something people would use?)

TheNailDev commented:
I just changed `TSong.Mp3` to `TSong.Audio`.
I did not notice any unexpected side effects.

I also changed the English and German localization to use Audio instead of MP3.

barbeque-squared commented on .gitignore:
```diff
@@ -48,7 +48,7 @@ ppaslink.sh
 
 # dev files
-/backup/
+**/backup/
```
This should be reverted. I'm totally on board with reevaluating whether or not this is useful or not in a separate ticket/PR, but it has nothing to do with this functionality, and the commit message also doesn't make it clear why this needed to be changed.

(it's possible it was just committed by accident, but even if everyone agrees it's an improvement, this isn't the place to make such a change)

TheNailDev replied:
My bad :sweat_smile: 
Reverted it

barbeque-squared commented:
re: https://github.com/UltraStar-Deluxe/USDX/pull/951#issuecomment-2606707094

Yeah independent (or at least _multiple_) PRs is the way to go for that. But can we _release_ this without all the other things? Personally I'm fine with partial support as long as it doesn't really break the not-yet implemented bits, but I'm not an end user.

Also generally agree with your comments on the subfeatures.

TheNailDev commented:
> But can we release this without all the other things?

In my opinion yes. As long as we don't say that we fully support 1.1.0 there should be no Issues.

For songs with Format < 1.1.0 nothing changes.

For songs with Format >= 1.1.0 the AUDIO and MP3 tag should (hopefully) both be valid audio files (usually the same file).
So for those songs also nothing would change. (Unless AUDIO is an invalid file, while MP3 is a valid file. But then I would say that the user should have the responsibility of having valid files and this should not be our issue)

barbeque-squared reviewed (approved):
Code looks good. Verified that existing songs still work. Did not actually test the new header, but the code looks good enough. Will merge tomorrow.
